docs/code/IDataViewDesignPrinciples.md

```diff
@@ -0,0 +1,479 @@
+  [here](#batch-parallel-cursoring).
+
+* **Large data support**: Constructing views on data files and cursoring
+  through the rows of a view does not require the entire data to fit in
+  memory. Conversely, when the entire data fits, there is nothing preventing
+  it from being loaded entirely in memory. See [here](#data-size).
+
+### Design Non-requirements
+
+The IDataView system design does *not* include the following:
+
+* **Multi-view schema information**: There is no direct support for specifying
+  cross-view schema information, for example, that certain columns are primary
+  keys, and that there are foreign key relationships among tables. However,
+  the column metadata support, together with conventions, may be used to
+  represent such information.
+
+* **Standard ML schema**: The IDataView system does not define, nor prescribe,
+  standard ML schema representation. For example, it does not dictate
+  representation of nor distinction between different semantic interpretations
+  of columns, such as label, feature, score, weight, etc. However, the column
+  metadata support, together with conventions, may be used to represent such
+  interpretations.
+
+* **Row count**: A view is not required to provide its row count. The
+  `IDataView` interface has a `GetRowCount` method with type `Nullable<long>`.
+  When this returns `null`, the row count is not available directly from the
+  view.
+
+* **Efficient indexed row access**: There is no standard way in the IDataView
+  system to request the values for a specific row number. While the
+  `IRowCursor` interface has a `MoveMany(long count)` method, it only supports
+  moving forward `(count > 0)`, and is not necessarily more efficient than
+  calling `MoveNext()` repeatedly. See [here](#row-cursor).
+
+* **Data file formats**: The IDataView system does not dictate storage or
+  transport formats. It *does* include interfaces for loader and saver
+  components. The AzureML Algorithms team has implemented loaders and savers
+  for some binary and text file formats, but additional loaders and savers can
+  (and will) be implemented. In particular, implementing a loader from XDF
+  will be straightforward. Implementing a saver to XDF will likely require the
+  XDF format to be extended to support vector-valued columns.
+
+* **Multi-node computation over multiple data partitions**: The IDataView
+  design is focused on single node computation. We expect that in multi-node
+  applications, each node will be given its own data partition(s) to operate
+  on, with aggregation happening outside an IDataView pipeline.
+
+## Schema and Type System
+
+### Basics
+
+IDataView has general schema support, in that a view can have an arbitrary
+number of columns, each having an associated name, index, data type, and
+optional metadata.
+
+Column names are case sensitive. Multiple columns can share the same name, in
+which case, one of the columns hides the others, in the sense that the name
```

Comment:
Hmmm. While this is part of the *implementations*, it isn't really part of the spec or `ISchema` interface. If you were to construct a dataview with columns `["A", "B", "B"]`, but the schema implementation behaves so that `TryGetColumnIndex("B", out int colIdx)" succeeds with `colIndex==1`, it would behave perfectly in, say, a transform pipeline and whatnot. I will though add a note about this in the implementation -- actually maybe a bit more, since I see that I had neglected to mention anything about the schema in the implementation notes aside from one or two notes about metadata...

---
In reply to: [188891796](https://github.com/dotnet/machinelearning/pull/173#discussion_r188891796) [](ancestors = 188891796)